### PR TITLE
Basic support for regular expression complement

### LINKED ISSUE #3408: Implement re.complement
We should support the regular expression complement operator `re.complement` .

### PR DESCRIPTION
**Pull request by ajreynol**
Fixes #3408 .

Adds basic rewriter, parsing, type checking and implements the required cases in regexp_operation.cpp.  It also adds some missing documentation in regexp_operation.h

### CODE REVIEW

File: src/theory/strings/regexp_operation.h [hunk 1]
```diff
@@ -41,7 +41,9 @@ namespace strings {
  */
 enum RegExpConstType
 {
-  // the regular expression doesn't contain variables or re.allchar or re.range
+  // the regular expression doesn't contain variables or re.complement or
+  // re.allchar or re.range. Notice that re.complement is grouped in this
+  // case since it can be seen as indirectly defined in terms of re.allchar.
```
Comment (4tXJ7f) [inline src/theory/strings/regexp_operation.h]: If we keep `re.complement` together with `re.allchar` and `re.range` then the last sentence should be moved to the comment for `RE_C_CONSTANT`.

File: src/theory/strings/regexp_operation.cpp [hunk 2]
```diff
@@ -264,6 +266,13 @@ int RegExpOpr::delta( Node r, Node &exp ) {
         }
         break;
       }
+      case kind::REGEXP_COMPLEMENT:
+      {
+        int tmp = delta(r[0], exp);
+        // flip the result if known
+        tmp = tmp == 0 ? 0 : (3 - tmp);
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/strings/regexp_operation.cpp]: Yes, good catch.

File: src/theory/strings/regexp_operation.cpp [hunk 3]
```diff
@@ -1222,6 +1245,12 @@ void RegExpOpr::simplifyPRegExp( Node s, Node r, std::vector< Node > &new_nodes
         }
         break;
       }
+      case kind::REGEXP_COMPLEMENT:
+      {
+        // s in complement(R) ---> ~( s in R )
+        conc = nm->mkNode(STRING_IN_REGEXP, s, r[0]).negate();
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/strings/regexp_operation.cpp]: This rule is in the rewriter. My understanding is that this function might be called on non-rewritten nodes (or even if it isn't then this case is just here just to be safe).

File: src/theory/strings/regexp_operation.cpp [hunk 4]
```diff
@@ -1046,6 +1063,12 @@ void RegExpOpr::simplifyNRegExp( Node s, Node r, std::vector< Node > &new_nodes
         }
         break;
       }
+      case kind::REGEXP_COMPLEMENT:
+      {
+        // ~( s in complement(R) ) ---> s in R
+        conc = nm->mkNode(STRING_IN_REGEXP, s, r[0]);
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/strings/regexp_operation.cpp]: Ditto.

File: src/theory/strings/regexp_operation.cpp [hunk 5]
```diff
@@ -77,7 +77,8 @@ RegExpConstType RegExpOpr::getRegExpConstType(Node r)
         d_constCache[cur] =
             tmp.isConst() ? RE_C_CONRETE_CONSTANT : RE_C_VARIABLE;
       }
-      else if (ck == REGEXP_SIGMA || ck == REGEXP_RANGE)
+      else if (ck == REGEXP_SIGMA || ck == REGEXP_RANGE
+               || ck == REGEXP_COMPLEMENT)
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/strings/regexp_operation.cpp]: I think I'd keep it like this. The intuition here is that regular expression complement is a regular expression operator, but is one that may be expensive to deal with in e.g. regular expression intersection.

**Comment by ajreynol** [comment]
Fixes #10 from cvc4-projects.

**Comment by ajreynol** [comment]
Currently blocking support for string standard.

File: src/theory/strings/regexp_operation.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/strings/regexp_operation.cpp]: I am still not a big fan of this: Note that we consider `RE_C_VARIABLE` to be "more expensive" than `RE_C_CONSTANT`. My reasoning is that we can have `re.comp(str.to.re(x))` where `x` is a string variable. We could also treat `re.comp` as the same class as its regular expression argument.

File: src/theory/strings/regexp_operation.h (hunk 1 above)
Comment (4tXJ7f) [reply src/theory/strings/regexp_operation.h]: This comment still needs to be updated.

File: src/theory/strings/regexp_operation.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/strings/regexp_operation.cpp]: Yes, good catch. It was intended to do the traversal to discover a case like this, I've fixed this in my latest commit.

**Comment by ajreynol** [comment]
Let me give a convincing argument why the re.comp is `C_CONSTANT` (after the traversal fix):  (re.comp R) is equivalent to (re.diff re.allchar R), hence it implicitly contains re.allchar. The comment above the enum reflects this.

**Pull request merged**